---
title: "portsmouth-darek-mdxbrain"
output:
  html_document:
    toc: yes
    df_print: paged

date: "_Last draft: `r format(Sys.time(), '%d %B, %Y %H:%M:%S')`_"
description: "mdx mice analysis"

knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_file = paste0(
        xfun::sans_ext(input), '-', format(Sys.time(), "%Y-%m-%d_%H-%M"), '.html'
      ),
      knit_root_dir = xfun::proj_root(),
      envir = globalenv()
    )
  })
---


```{r install_libraries, eval=FALSE, include=FALSE}
library(devtools)
devtools::install_github("zhangyuqing/sva-devel")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
# https://github.com/Bioconductor/bioconductor_docker/issues/22
BiocManager::install("preprocessCore", configure.args="--disable-threading")
```

```{r load_libraries, include=FALSE}
library(tidyverse)
library(preprocessCore)
library(RColorBrewer)

source('analysis/utils.R')
library(edgeR)
```

```{r read_data}
sample_info_rds_path <- 'data/sample_info.RDS'
sample_info <- readRDS(sample_info_rds_path)

counts_cerebrum_raw <-
  read_tsv('raw/cerebrum_gene-level-counts-annotated.tsv') %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  )
counts_cerebellum_raw <-
  read_tsv('raw/cerebellum_gene-level-counts-annotated.tsv') %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  )
counts_raw <-
  full_join(
    read_tsv('raw/cerebrum_gene-level-counts-annotated.tsv'),
    read_tsv('raw/cerebellum_gene-level-counts-annotated.tsv'),
    by = c(
      'Gene stable ID', 'Transcript stable ID', 'Exon stable ID',
      'Gene description', 'Chromosome/scaffold name', 'Gene start (bp)',
      'Gene end (bp)', 'Gene name', 'Gene type', 'Gene Synonym'
    )
  ) %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  )
```

```{r prepare_data}
counts_cerebrum_normalized_log <-
  counts_cerebrum_raw %>%
  select(starts_with('sample_')) %>%
  as.matrix() %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}
counts_cerebellum_normalized_log <-
  counts_cerebellum_raw %>%
  select(starts_with('sample_')) %>%
  as.matrix() %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}
counts_normalized_log <-
  counts_raw %>%
  select(starts_with('sample_')) %>%
  as.matrix() %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}

counts <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_normalized_log)
counts_cerebrum <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_cerebrum_normalized_log)
counts_cerebellum <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_cerebellum_normalized_log)
```

```{r}
plot_info <-
  sample_info %>%
  arrange(age_days, desc(tissue), desc(genotype)) %>%
  mutate(
    `wt / mdx` = if_else(genotype == 'wt', 'yellow', 'black'),
    `cerebrum / cerebellum` = if_else(tissue == 'cerebrum', 'red', 'blue'),
    `10 days / 70 days` = if_else(age_days == 10, 'green', 'magenta'),
  )
```

```{r fit_model}
edger_model <- function(x, sample_info_groups, tissue=NA, contrast='genotype'){
  mdx_groups. <-
    sample_info_groups %>%
    `[`(match(colnames(x), .$id), ) %>%
    pull(group) %>%
    factor()

  dge_counts <- DGEList(counts = x, group = mdx_groups.)
  keep_counts <- filterByExpr(dge_counts)
  dge_counts <- dge_counts[keep_counts, , keep.lib.sizes = FALSE]

  dge_counts <- calcNormFactors(dge_counts)
  design <- model.matrix(~ 0 + mdx_groups.)
  dge_counts <- estimateDisp(dge_counts, design)

  fit <- glmQLFit(dge_counts, design)

  # contrasts
  if (tissue %in% c('cerebrum', 'cerebellum')){
    contrasts <- makeContrasts(
      # overall genotype effect
      genotype=(
        (mdx_groups.wt_10 + mdx_groups.wt_70)
        - (mdx_groups.mdx_10 + mdx_groups.mdx_70)
      ),
      genotype_10=mdx_groups.wt_10 - mdx_groups.mdx_10,
      genotype_70=mdx_groups.wt_70 - mdx_groups.mdx_70,
      levels = design
    )
  } else {
    contrasts <- makeContrasts(
      # overall genotype effect
      genotype=
        (
          mdx_groups.wt_10_cerebrum + mdx_groups.wt_70_cerebrum
          + mdx_groups.wt_10_cerebellum + mdx_groups.wt_70_cerebellum
        )
        - (
          mdx_groups.mdx_10_cerebrum + mdx_groups.mdx_70_cerebrum
          + mdx_groups.mdx_10_cerebellum + mdx_groups.mdx_70_cerebellum
        ),
      levels = design
    )
  }
  results <- glmQLFTest(fit, contrast=contrasts[, contrast])

  topTags(results, n = 'inf')$table %>%
    as_tibble(rownames = 'row_index')
}
```


# All tissues

```{r}
x <- select(counts_raw, matches('sample_'))
sample_info_groups <-
  sample_info %>%
    mutate(group = paste(genotype, age_days, tissue, sep = '_'))

edger_genes <-
  edger_model(x, sample_info_groups) %>%
  mutate(
    gene_name = counts_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_raw[as.numeric(.$row_index), ]$gene_stable_id,
    .before = everything()
  ) %>%
  select(-row_index)
```

```{r draw_heatmap_03, echo=FALSE, fig.align="center", fig.height=10, fig.width=10, results='asis', include=TRUE, warning=FALSE}
plot_data <-
  counts %>%
  full_join(edger_genes, by = c('gene_stable_id', 'gene_name')) %>%
  filter(FDR < 0.001)

.x <- plot_data %>% select(all_of(plot_info$id)) %>% as.matrix()
rownames(.x) <- plot_data$gene_name
.rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
.col_side_colors <- plot_info %>% select(contains('/')) %>% as.matrix()
.way = 'genotype / genotype:tissue / genotype:age_days / genotype:tissue:age_days'

h <- heatmap.3(
    x = .x,
    main = 'edgeR',
    Rowv = TRUE,
    Colv = FALSE,
    dendrogram = 'row',
    distfun = function(x) as.dist(1-cor(t(x))),
    scale = "row",
    breaks = seq(-2.5, 2.5, length.out = 25),
    col = rev(.rdbu_ramp(24)),
    trace = "none",
    ColSideColors = .col_side_colors,
    ColSideColorsSize = 6,
    key = FALSE,
    lhei = c(1.5, 20),
    lwid = c(1,5),
    cexRow = 1.2,
    cexCol = 1.5,
    margins = c(9, 8),
)
```

```{r}
ct <- cutree(as.hclust(h$rowDendrogram), k = 4)
ct_heatmap_order <- ct[rev(h$rowInd)]

clusters <- tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order)

clusters %>%
  filter(cluster != 0) %>%
  pull(gene_name) %>%
  paste(collapse = '\n') %>%
  cat
```


# Cerebrum

```{r edger_cerebrum}
tiss = 'cerebrum'

plot_info_cerebrum <-
  plot_info %>%
  filter(tissue == tiss) %>%
  mutate(
    `cerebrum /` = `cerebrum / cerebellum`,
    .before = `cerebrum / cerebellum`,
    .keep = 'unused'
  )

x <- select(counts_cerebrum_raw, starts_with('sample_'))
sample_info_groups <-
  sample_info %>%
  filter(tissue == tiss) %>%
  mutate(group = paste(genotype, age_days, sep = '_'))

edger_genes <-
  edger_model(x, sample_info_groups, tiss) %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    .before = everything()
  ) %>%
  select(-row_index)
```

```{r draw_heatmap_03, echo=FALSE, fig.align="center", fig.height=12, fig.width=8, results='asis', include=TRUE, warning=FALSE}
fdr_th <- 0.00005

plot_data <-
  counts_cerebrum %>%
  full_join(edger_genes, by = c('gene_stable_id', 'gene_name')) %>%
  filter(FDR < fdr_th) %>%
  filter(gene_type == 'protein_coding') %>%
  `[`(-duplicated, )

.x <- plot_data %>% select(any_of(plot_info_cerebrum$id)) %>% as.matrix()
rownames(.x) <- plot_data$gene_name
.rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
.col_side_colors <- plot_info_cerebrum %>% select(contains('/')) %>% as.matrix()
.way = 'genotype / genotype:tissue / genotype:age_days / genotype:tissue:age_days'

path <- paste0('results/genotype-edger-', tiss, '.png')
png(path, width = 1000*2.0, height = 2200*2.0, pointsize = 25)
h_cerebrum <- heatmap.3(
    x = .x,
    main = paste('genotype effect in ', tiss, ' FDR <', format(fdr_th, scientific = FALSE)),
    Rowv = TRUE,
    Colv = FALSE,
    dendrogram = 'row',
    distfun = function(x) as.dist(1-cor(t(x))),
    scale = "row",
    breaks = seq(-2.5, 2.5, length.out = 25),
    col = rev(.rdbu_ramp(24)),
    trace = "none",
    ColSideColors = .col_side_colors,
    ColSideColorsSize = 5,
    key = FALSE,
    lhei = c(1.0, 40),
    lwid = c(1,5),
    cexRow = 1.3,
    cexCol = 1.3,
    margins = c(9, 8),
)
dev.off()
```

```{r}
ct <- cutree(as.hclust(h_cerebrum$rowDendrogram), h = 0.002)
tibble(
  cluster=ct,
  gene_name=names(ct),
  m=h_cerebrum$rowMeans
) %>%
group_by(cluster) %>%
filter(n() > 1) %>%
arrange(-m)

duplicated <- which((ct == 15) & names(ct) != 'Pcdha4')
```

```{r}
ct <- cutree(as.hclust(h_cerebrum$rowDendrogram), k = 4)
ct_heatmap_order <- ct[rev(h_cerebrum$rowInd)]

# tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order) %>%
#   filter(cluster != 0) %>%
#   pull(gene_name) %>%
#   paste(collapse = '\n') %>%
#   cat

tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order) %>%
  write_csv(paste0('results/edger-', tiss, '-clusters.csv'))
```


# Cerebellum

```{r edger_cerebellum}
tiss = 'cerebellum'

plot_info_cerebellum <-
  plot_info %>%
  filter(tissue == tiss) %>%
  mutate(
    `/ cerebellum` = `cerebrum / cerebellum`,
    .before = `cerebrum / cerebellum`,
    .keep = 'unused'
  )

x <- select(counts_cerebellum_raw, starts_with('sample_'))
sample_info_groups <-
  sample_info %>%
  filter(tissue == tiss) %>%
  mutate(group = paste(genotype, age_days, sep = '_'))

edger_genes <-
  edger_model(x, sample_info_groups, tiss) %>%
  mutate(
    gene_name = counts_cerebellum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebellum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    .before = everything()
  ) %>%
  select(-row_index)
```

```{r draw_heatmap_03, echo=FALSE, fig.align="center", fig.height=13, fig.width=8, results='asis', include=TRUE, warning=FALSE}
fdr_th <- 0.1

plot_data <-
  counts_cerebellum %>%
  full_join(edger_genes, by = c('gene_stable_id', 'gene_name')) %>%
  filter(FDR < fdr_th) %>%
  filter(gene_type == 'protein_coding')

.x <- plot_data %>% select(any_of(plot_info_cerebellum$id)) %>% as.matrix()
rownames(.x) <- plot_data$gene_name
.rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
.col_side_colors <- plot_info_cerebellum %>% select(contains('/')) %>% as.matrix()
.way = 'genotype / genotype:tissue / genotype:age_days / genotype:tissue:age_days'

path <- paste0('results/genotype-edger-', tiss, '.png')
png(path, width = 1000*1.0, height = 1800*1.0, pointsize = 20)
h_cerebellum <- heatmap.3(
    x = .x,
    main = paste('genotype effect in ', tiss, 'FDR <', format(fdr_th, scientific = FALSE)),
    Rowv = TRUE,
    Colv = FALSE,
    dendrogram = 'row',
    distfun = function(x) as.dist(1-cor(t(x))),
    scale = "row",
    breaks = seq(-2.5, 2.5, length.out = 25),
    col = rev(.rdbu_ramp(24)),
    trace = "none",
    ColSideColors = .col_side_colors,
    ColSideColorsSize = 6,
    key = FALSE,
    lhei = c(1.5, 20),
    lwid = c(1,5),
    cexRow = 1.3,
    cexCol = 1.3,
    margins = c(9, 8),
)
dev.off()
```

```{r}
ct <- cutree(as.hclust(h_cerebellum$rowDendrogram), k = 4)
ct_heatmap_order <- ct[rev(h_cerebellum$rowInd)]

# tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order) %>%
#   filter(cluster != 0) %>%
#   pull(gene_name) %>%
#   paste(collapse = '\n') %>%
#   cat

tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order) %>%
  write_csv(paste0('results/edger-', tiss, '-clusters.csv'))
```
