---
title: "portsmouth-darek-mdxbrain"
output:
  html_document:
    toc: yes
    df_print: paged

date: "_Last draft: `r format(Sys.time(), '%d %B, %Y %H:%M:%S')`_"
description: "mdx mice analysis"

knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_file = paste0(
        xfun::sans_ext(input), '-', format(Sys.time(), "%Y-%m-%d_%H-%M"), '.html'
      ),
      knit_root_dir = xfun::proj_root(),
      envir = globalenv()
    )
  })
---


```{r install_libraries, eval=FALSE, include=FALSE}
library(devtools)
devtools::install_github("zhangyuqing/sva-devel")
devtools::install_github("yanlinlin82/ggvenn")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
# https://github.com/Bioconductor/bioconductor_docker/issues/22
BiocManager::install("preprocessCore", configure.args="--disable-threading")
```

```{r load_libraries, include=FALSE}
library(IRanges)
library(sets)
library(preprocessCore)
library(RColorBrewer)
library(tidyverse)

source('analysis/utils.R')
library(edgeR)
library(ggvenn)
```

```{r}
expression_dist_matrix <- function(counts_df) {
  counts_df %>%
  select(starts_with('sample_')) %>%
  as.matrix() %>%
  # library(Rfast2)
  # t() %>%
  # dcora()  # correlation as distance
  dist(method = 'canberra') %>%
  as.matrix()
}

overlap_ratio_matrix <- function(counts_df){
  query <- IRanges(
    counts_df$start,
    counts_df$end
  )
  
  overlap_pairs <-
    findOverlaps(query) %>%
    as.data.frame() %>%
    mutate(
      query_start = counts_df$start[queryHits],
      query_end = counts_df$end[queryHits],
      subject_start = counts_df$start[subjectHits],
      subject_end = counts_df$end[subjectHits],
      query_gene_name = counts_df$gene_name[queryHits],
      subject_gene_name = counts_df$gene_name[subjectHits],
      query_chr = counts_df$chr[queryHits],
      subject_chr = counts_df$chr[subjectHits],
    ) %>%
    mutate(
      overlap_width = pmin(query_end, subject_end)
      - pmax(query_start, subject_start),
      shorter_range = pmin(query_end - query_start, subject_end - subject_start),
      longer_range = pmax(query_end - query_start, subject_end - subject_start),
    ) %>%
    mutate(
      # ratio = overlap_width / longer_range
      ratio = if_else(query_chr == subject_chr, overlap_width / shorter_range, 0)
    )
    
  overlap_matrix <- matrix(0, nrow = nrow(dist_matrix), ncol = ncol(dist_matrix))
  for (i in 1:nrow(overlap_pairs)) {
    query <- overlap_pairs[i, 'queryHits']
    subject <- overlap_pairs[i, 'subjectHits']
    ratio <- overlap_pairs[i, 'ratio']
  
    overlap_matrix[query, subject] <- ratio
  }
  overlap_matrix
}


reduce_sets <- function(gene_sets) {
  out <- list(gene_sets[[1]])
  for (x in gene_sets) {
    create_new <- TRUE
    for (iy in 1:length(out)) {
      y <- out[[iy]]
      if (!set_is_empty(set_intersection(x, y))) {
        out[[iy]] <- sort(unique(c(out[[iy]], x)))
        create_new <- FALSE
      }
    }
    if (create_new)
      out[[length(out) + 1]] <- x
  }
  out
}

gene_sets <- function(m) {
  gene_sets <- apply(unname(m), 1, which)
  new_gene_sets <- list()
  for (x in gene_sets) {
    x <- unname(x)
    if (length(x) > 2) {
      new_gene_sets[[length(new_gene_sets) + 1]] <- x
    }
  }
  
  if (length(new_gene_sets) == 0) {
    return()
  }
  
  old_gene_sets <- list()
  while(!identical(old_gene_sets, new_gene_sets)) {
    old_gene_sets <- new_gene_sets
    new_gene_sets <- reduce_sets(old_gene_sets)
  }
  new_gene_sets
}

```

```{r read_data}
sample_info_rds_path <- 'data/sample_info.RDS'
sample_info <- readRDS(sample_info_rds_path)

counts_cerebrum_raw <-
  read_tsv('raw/cerebrum_gene-level-counts-annotated.tsv') %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  )
counts_cerebellum_raw <-
  read_tsv('raw/cerebellum_gene-level-counts-annotated.tsv') %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  )
counts_raw <-
  full_join(
    read_tsv('raw/cerebrum_gene-level-counts-annotated.tsv'),
    read_tsv('raw/cerebellum_gene-level-counts-annotated.tsv'),
    by = c(
      'Gene stable ID', 'Transcript stable ID', 'Exon stable ID',
      'Gene description', 'Chromosome/scaffold name', 'Gene start (bp)',
      'Gene end (bp)', 'Gene name', 'Gene type', 'Gene Synonym'
    )
  ) %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  )
```

```{r remove_duplicated_genes}
duplicated_genes <- c()

chr_subset <- c('1', '2', '3', '4', '5', '6', '7', '8')
counts_chr <-
  counts_raw %>%
  filter(gene_type == 'protein_coding') %>%
  rename(start=`Gene start (bp)`, end=`Gene end (bp)`) %>%
  select(gene_stable_id, gene_name, chr, start, end, starts_with('sample_')) %>%
  mutate(g = chr %in% chr_subset)
  # mutate(dup = duplicated(gene_name))

dist_matrix <- expression_dist_matrix(filter(counts_chr, g))
overlap_matrix <- overlap_ratio_matrix(filter(counts_chr, g))
result <- (overlap_matrix > 0.5) & (dist_matrix < 1.0)

for (s in gene_sets(result)) {
  print(s)
  genes <- 
    counts_chr %>%
    filter(g) %>%
    slice(s) %>%
    mutate(sum = apply(select(., starts_with('sample_')), 1, sum)) %>%
    arrange(-sum) %>%
    slice(-1) %>%
    pull(gene_stable_id)
  
  
    counts_chr %>%
    filter(g) %>%
    slice(s) %>%print
  
  duplicated_genes <- c(duplicated_genes, genes)
}

dist_matrix <- expression_dist_matrix(filter(counts_chr, !g))
overlap_matrix <- overlap_ratio_matrix(filter(counts_chr, !g))
result <- (overlap_matrix > 0.5) & (dist_matrix < 1)

for (s in gene_sets(result)) {
  print(s)
  genes <- 
    counts_chr %>%
    filter(!g) %>%
    slice(s) %>%
    mutate(sum = apply(select(., starts_with('sample_')), 1, sum)) %>%
    arrange(-sum) %>%
    slice(-1) %>%
    pull(gene_stable_id)

    counts_chr %>%
    filter(!g) %>%
    slice(s) %>%print
  
  duplicated_genes <- c(duplicated_genes, genes)
}

counts_cerebrum_raw <-
  filter(counts_cerebrum_raw, !gene_stable_id %in% duplicated_genes)
counts_cerebellum_raw <-
  filter(counts_cerebellum_raw, !gene_stable_id %in% duplicated_genes)
counts_raw <-
  filter(counts_raw, !gene_stable_id %in% duplicated_genes)
```

```{r prepare_data}
counts_cerebrum_normalized_log <-
  counts_cerebrum_raw %>%
  select(starts_with('sample_')) %>%
  as.matrix() %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}
counts_cerebellum_normalized_log <-
  counts_cerebellum_raw %>%
  select(starts_with('sample_')) %>%
  as.matrix() %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}
counts_normalized_log <-
  counts_raw %>%
  select(starts_with('sample_')) %>%
  as.matrix() %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}

counts <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_normalized_log)
counts_cerebrum <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_cerebrum_normalized_log)
counts_cerebellum <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_cerebellum_normalized_log)
```

```{r}
plot_info <-
  sample_info %>%
  arrange(age_days, desc(tissue), desc(genotype)) %>%
  mutate(
    `wt / mdx` = if_else(genotype == 'wt', 'yellow', 'black'),
    `cerebrum / cerebellum` = if_else(tissue == 'cerebrum', 'red', 'blue'),
    `10 days / 70 days` = if_else(age_days == 10, 'green', 'magenta'),
  )
```

```{r fit_model}
edger_model <- function(x, sample_info_groups, tissue=NA, contrast='genotype'){
  mdx_groups. <-
    sample_info_groups %>%
    `[`(match(colnames(x), .$id), ) %>%
    pull(group) %>%
    factor()

  dge_counts <- DGEList(counts = x, group = mdx_groups.)
  keep_counts <- filterByExpr(dge_counts)
  dge_counts <- dge_counts[keep_counts, , keep.lib.sizes = FALSE]

  dge_counts <- calcNormFactors(dge_counts)
  design <- model.matrix(~ 0 + mdx_groups.)
  dge_counts <- estimateDisp(dge_counts, design)

  fit <- glmQLFit(dge_counts, design)

  # contrasts
  if (tissue %in% c('cerebrum', 'cerebellum')){
    contrasts <- makeContrasts(
      # overall genotype effect
      genotype=(
        (mdx_groups.wt_10 + mdx_groups.wt_70)
        - (mdx_groups.mdx_10 + mdx_groups.mdx_70)
      ),
      genotype_10=mdx_groups.wt_10 - mdx_groups.mdx_10,
      genotype_70=mdx_groups.wt_70 - mdx_groups.mdx_70,
      age_days=(
        (mdx_groups.wt_10 + mdx_groups.mdx_10)
        - (mdx_groups.wt_70 + mdx_groups.mdx_70)
      ),
      levels = design
    )
  } else {
    contrasts <- makeContrasts(
      # overall genotype effect
      genotype=
        (
          mdx_groups.wt_10_cerebrum + mdx_groups.wt_70_cerebrum
          + mdx_groups.wt_10_cerebellum + mdx_groups.wt_70_cerebellum
        )
        - (
          mdx_groups.mdx_10_cerebrum + mdx_groups.mdx_70_cerebrum
          + mdx_groups.mdx_10_cerebellum + mdx_groups.mdx_70_cerebellum
        ),
      levels = design
    )
  }
  results <- glmQLFTest(fit, contrast=contrasts[, contrast])

  topTags(results, n = 'inf')$table %>%
    as_tibble(rownames = 'row_index')
}
```


# All tissues

```{r}
x <- select(counts_raw, matches('sample_'))
sample_info_groups <-
  sample_info %>%
    mutate(group = paste(genotype, age_days, tissue, sep = '_'))

edger_genes <-
  edger_model(x, sample_info_groups) %>%
  mutate(
    gene_name = counts_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_raw[as.numeric(.$row_index), ]$gene_stable_id,
    .before = everything()
  ) %>%
  select(-row_index)
```

```{r draw_heatmap_03, echo=FALSE, fig.align="center", fig.height=10, fig.width=10, results='asis', include=TRUE, warning=FALSE}
plot_data <-
  counts %>%
  full_join(edger_genes, by = c('gene_stable_id', 'gene_name')) %>%
  filter(FDR < 0.001)

.x <- plot_data %>% select(all_of(plot_info$id)) %>% as.matrix()
rownames(.x) <- plot_data$gene_name
.rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
.col_side_colors <- plot_info %>% select(contains('/')) %>% as.matrix()
.way = 'genotype / genotype:tissue / genotype:age_days / genotype:tissue:age_days'

h <- heatmap.3(
    x = .x,
    main = 'edgeR',
    Rowv = TRUE,
    Colv = FALSE,
    dendrogram = 'row',
    distfun = function(x) as.dist(1-cor(t(x))),
    scale = "row",
    breaks = seq(-2.5, 2.5, length.out = 25),
    col = rev(.rdbu_ramp(24)),
    trace = "none",
    ColSideColors = .col_side_colors,
    ColSideColorsSize = 6,
    key = FALSE,
    lhei = c(1.5, 20),
    lwid = c(1,5),
    cexRow = 1.2,
    cexCol = 1.5,
    margins = c(9, 8),
)
```

```{r}
ct <- cutree(as.hclust(h$rowDendrogram), k = 4)
ct_heatmap_order <- ct[rev(h$rowInd)]

clusters <- tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order)

clusters %>%
  filter(cluster != 0) %>%
  pull(gene_name) %>%
  paste(collapse = '\n') %>%
  cat
```


# Cerebrum

```{r edger_cerebrum}
tiss = 'cerebrum'

plot_info_cerebrum <-
  plot_info %>%
  filter(tissue == tiss) %>%
  mutate(
    `cerebrum /` = `cerebrum / cerebellum`,
    .before = `cerebrum / cerebellum`,
    .keep = 'unused'
  )

x <- select(counts_cerebrum_raw, starts_with('sample_'))
sample_info_groups <-
  sample_info %>%
  filter(tissue == tiss) %>%
  mutate(group = paste(genotype, age_days, sep = '_'))

edger_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'genotype') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    .before = everything()
  ) %>%
  select(-row_index)
```

```{r draw_heatmap_03, echo=FALSE, fig.align="center", fig.height=12, fig.width=8, results='asis', include=TRUE, warning=FALSE}
fdr_th <- 0.00005

plot_data <-
  counts_cerebrum %>%
  full_join(edger_genes, by = c('gene_stable_id', 'gene_name')) %>%
  filter(FDR < fdr_th) %>%
  filter(gene_type == 'protein_coding') %>%
  `[`(-duplicated, )

.x <- plot_data %>% select(any_of(plot_info_cerebrum$id)) %>% as.matrix()
rownames(.x) <- plot_data$gene_name
colnames(.x) <-
  tibble(id = colnames(.x)) %>%
  left_join(sample_info, by = 'id') %>%
  mutate(age_days = if_else(age_days == 10, '10 days', '10 weeks')) %>%
  unite('col_names', genotype, age_days, sep = ' ') %>%
  pull(col_names)
.rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
.col_side_colors <- plot_info_cerebrum %>% select(contains('/'), -contains(tiss)) %>% as.matrix()
.way = 'genotype / genotype:tissue / genotype:age_days / genotype:tissue:age_days'
.col_sep <- 1:3 * 5

path <- paste0('results/genotype-edger-', tiss, '.svg')
svg(path, width = 1000*0.03, height = 2200*0.03, pointsize = 25)
par(cex.axis=1.3)
h_cerebrum <- heatmap.3(
    x = .x,
    main = paste('genotype effect in ', tiss, ' FDR <', format(fdr_th, scientific = FALSE)),
    Rowv = TRUE,
    Colv = FALSE,
    dendrogram = 'row',
    distfun = function(x) as.dist(1-cor(t(x))),
    scale = "row",
    breaks = seq(-2.5, 2.5, length.out = 25),
    col = rev(.rdbu_ramp(24)),
    trace = "none",
    ColSideColors = .col_side_colors,
    ColSideColorsSize = 5,
    colsep = .col_sep,
    srtCol = 45,  # doesn't work in heatmap.3
    key = FALSE,
    lhei = c(1.0, 40),
    lwid = c(1,5),
    cexRow = 1.3,
    cexCol = 1.3,
    margins = c(10, 8),
)
dev.off()
```

```{r}
ct <- cutree(as.hclust(h_cerebrum$rowDendrogram), h = 0.002)
tibble(
  cluster=ct,
  gene_name=names(ct),
  m=h_cerebrum$rowMeans
) %>%
group_by(cluster) %>%
filter(n() > 1) %>%
arrange(-m)

duplicated <- which((ct == 15) & names(ct) != 'Pcdha4')
```

```{r}
ct <- cutree(as.hclust(h_cerebrum$rowDendrogram), k = 4)
ct_heatmap_order <- ct[rev(h_cerebrum$rowInd)]

# tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order) %>%
#   filter(cluster != 0) %>%
#   pull(gene_name) %>%
#   paste(collapse = '\n') %>%
#   cat

tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order) %>%
  write_csv(paste0('results/edger-', tiss, '-clusters.csv'))
```

```{r venn}
th <- 0.1
venn_age_days_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'age_days') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    gene_type = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_type,
    .before = everything()
  ) %>%
  filter(gene_type == 'protein_coding') %>%
  select(-row_index)

venn_genotype_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'genotype') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    gene_type = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_type,
    .before = everything()
  ) %>%
  filter(gene_type == 'protein_coding') %>%
  select(-row_index)

venn_plot_data <-
  full_join(
    venn_age_days_genes,
    venn_genotype_genes,
    by = c('gene_name', 'gene_stable_id'),
    suffix = c('.age_days', '.genotype')
  ) %>%
  mutate(
    age = FDR.age_days < th,
    genotype = FDR.genotype < th,
    age_regulation = if_else(logFC.age_days < 0, 'up', 'down'),
    genotype_regulation = if_else(logFC.genotype < 0, 'up', 'down'),
  )

.venn_age_name <- venn_plot_data %>%
  filter(age) %>%
  group_by(age_regulation) %>%
  summarise(n = n()) %>%
  arrange(desc(age_regulation)) %>%
  mutate(name = paste(age_regulation, n)) %>%
  pull(name) %>%
  paste0(collapse = ' / ') %>%
  { paste0('age (', paste0(., collapse = ' / '), ')') }
.venn_genotype_name <- venn_plot_data %>%
  filter(genotype) %>%
  group_by(genotype_regulation) %>%
  summarise(n = n()) %>%
  arrange(desc(genotype_regulation)) %>%
  mutate(name = paste(genotype_regulation, n)) %>%
  pull(name) %>%
  { paste0('genotype (', paste0(., collapse = ' / '), ')') }

ggplot() +
  geom_text(
    mapping = aes(x = 0, y = 1.8),
    label = tiss,
    size = 6
  ) +
  geom_venn(
    aes_string(A = 'age', B = 'genotype'),
    data = venn_plot_data,
    fill_color = c('#0073C2FF', '#EFC000FF'),
    set_names = c(.venn_age_name, .venn_genotype_name),
    set_name_size = 4,
    auto_scale = TRUE,
    show_percentage = FALSE,
    show_outside = 'none',
  ) +
  labs(
    caption = paste('Number of genes regulated by age or genotype in', tiss, '; FDR <', th)
  ) +
  coord_fixed() +
  theme_void()
ggsave(
  paste0('results/venn-', tiss, '-', th, '.svg'),
  bg = 'white',
  width = 9,
  height = 6
)
```

```{r venn_10_70}
th <- 0.1
venn_genotype_10_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'genotype_10') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    gene_type = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_type,
    .before = everything()
  ) %>%
  filter(gene_type == 'protein_coding') %>%
  select(-row_index)

venn_genotype_70_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'genotype_70') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    gene_type = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_type,
    .before = everything()
  ) %>%
  filter(gene_type == 'protein_coding') %>%
  select(-row_index)


venn_plot_data <-
  full_join(
    venn_genotype_10_genes,
    venn_genotype_70_genes,
    by = c('gene_name', 'gene_stable_id'),
    suffix = c('.genotype_10', '.genotype_70')
  ) %>%
  mutate(
    genotype_10 = FDR.genotype_10 < th,
    genotype_70 = FDR.genotype_70 < th,
    genotype_10_regulation = if_else(logFC.genotype_10 < 0, 'up', 'down'),
    genotype_70_regulation = if_else(logFC.genotype_70 < 0, 'up', 'down'),
  )

.venn_genotype_10_name <- venn_plot_data %>%
  filter(genotype_10) %>%
  group_by(genotype_10_regulation) %>%
  summarise(n = n()) %>%
  arrange(desc(genotype_10_regulation)) %>%
  mutate(name = paste(genotype_10_regulation, n)) %>%
  pull(name) %>%
  paste0(collapse = ' / ') %>%
  { paste0('genotype 10 days (', paste0(., collapse = ' / '), ')') }
.venn_genotype_70_name <- venn_plot_data %>%
  filter(genotype_70) %>%
  group_by(genotype_70_regulation) %>%
  summarise(n = n()) %>%
  arrange(desc(genotype_70_regulation)) %>%
  mutate(name = paste(genotype_70_regulation, n)) %>%
  pull(name) %>%
  { paste0('genotype 10 weeks (', paste0(., collapse = ' / '), ')') }

ggplot() +
  geom_text(
    mapping = aes(x = 0, y = 1.8),
    label = tiss,
    size = 6
  ) +
  geom_venn(
    aes_string(A = 'genotype_10', B = 'genotype_70'),
    data = venn_plot_data,
    fill_color = c('#0073C2FF', '#EFC000FF'),
    set_names = c(.venn_genotype_10_name, .venn_genotype_70_name),
    set_name_size = 3,
    auto_scale = TRUE,
    show_percentage = FALSE,
    show_outside = 'none',
  ) +
  labs(
    caption = paste('Number of genes regulated by genotype in 10 or 70 days old mice in', tiss, '; FDR <', th)
  ) +
  coord_fixed() +
  theme_void()
ggsave(
  paste0('results/venn-', tiss, '-in-age-', th, '.svg'),
  bg = 'white',
  width = 9,
  height = 6
)
```



# Cerebellum

```{r edger_cerebellum}
tiss = 'cerebellum'

plot_info_cerebellum <-
  plot_info %>%
  filter(tissue == tiss) %>%
  mutate(
    `/ cerebellum` = `cerebrum / cerebellum`,
    .before = `cerebrum / cerebellum`,
    .keep = 'unused'
  )

x <- select(counts_cerebellum_raw, starts_with('sample_'))
sample_info_groups <-
  sample_info %>%
  filter(tissue == tiss) %>%
  mutate(group = paste(genotype, age_days, sep = '_'))

edger_genes <-
  edger_model(x, sample_info_groups, tiss) %>%
  mutate(
    gene_name = counts_cerebellum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebellum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    .before = everything()
  ) %>%
  select(-row_index)
```

```{r draw_heatmap_03, echo=FALSE, fig.align="center", fig.height=13, fig.width=8, results='asis', include=TRUE, warning=FALSE}
fdr_th <- 0.1

plot_data <-
  counts_cerebellum %>%
  full_join(edger_genes, by = c('gene_stable_id', 'gene_name')) %>%
  filter(FDR < fdr_th) %>%
  filter(gene_type == 'protein_coding')

.x <- plot_data %>% select(any_of(plot_info_cerebellum$id)) %>% as.matrix()
rownames(.x) <- plot_data$gene_name
colnames(.x) <-
  tibble(id = colnames(.x)) %>%
  left_join(sample_info, by = 'id') %>%
  mutate(age_days = if_else(age_days == 10, '10 days', '10 weeks')) %>%
  unite('col_names', genotype, age_days, sep = ' ') %>%
  pull(col_names)
.rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
.col_side_colors <- plot_info_cerebellum %>% select(contains('/'), -contains(tiss)) %>% as.matrix()
.col_sep <- 1:3 * 5
.way = 'genotype / genotype:tissue / genotype:age_days / genotype:tissue:age_days'

path <- paste0('results/genotype-edger-', tiss, '.svg')
svg(path, width = 1000*0.02, height = 1800*0.02, pointsize = 25)
par(cex.axis=1.3)
h_cerebellum <- heatmap.3(
    x = .x,
    main = paste('genotype effect in ', tiss, 'FDR <', format(fdr_th, scientific = FALSE)),
    Rowv = TRUE,
    Colv = FALSE,
    dendrogram = 'row',
    distfun = function(x) as.dist(1-cor(t(x))),
    scale = "row",
    breaks = seq(-2.5, 2.5, length.out = 25),
    col = rev(.rdbu_ramp(24)),
    trace = "none",
    ColSideColors = .col_side_colors,
    ColSideColorsSize = 5,
    colsep = .col_sep,
    key = FALSE,
    lhei = c(1.5, 20),
    lwid = c(1,5),
    cexRow = 1.3,
    cexCol = 1.3,
    margins = c(9, 8),
)
dev.off()
```

```{r}
ct <- cutree(as.hclust(h_cerebellum$rowDendrogram), k = 4)
ct_heatmap_order <- ct[rev(h_cerebellum$rowInd)]

# tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order) %>%
#   filter(cluster != 0) %>%
#   pull(gene_name) %>%
#   paste(collapse = '\n') %>%
#   cat

tibble(gene_name = names(ct_heatmap_order), cluster = ct_heatmap_order) %>%
  write_csv(paste0('results/edger-', tiss, '-clusters.csv'))
```

```{r venn2}
th <- 0.1
venn_age_days_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'age_days') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    gene_type = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_type,
    .before = everything()
  ) %>%
  filter(gene_type == 'protein_coding') %>%
  select(-row_index)

venn_genotype_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'genotype') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    gene_type = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_type,
    .before = everything()
  ) %>%
  filter(gene_type == 'protein_coding') %>%
  select(-row_index)

venn_plot_data <-
  full_join(
    venn_age_days_genes,
    venn_genotype_genes,
    by = c('gene_name', 'gene_stable_id'),
    suffix = c('.age_days', '.genotype')
  ) %>%
  mutate(
    age = FDR.age_days < th,
    genotype = FDR.genotype < th,
    age_regulation = if_else(logFC.age_days < 0, 'up', 'down'),
    genotype_regulation = if_else(logFC.genotype < 0, 'up', 'down'),
  )

.venn_age_name <- venn_plot_data %>%
  filter(age) %>%
  group_by(age_regulation) %>%
  summarise(n = n()) %>%
  arrange(desc(age_regulation)) %>%
  mutate(name = paste(age_regulation, n)) %>%
  pull(name) %>%
  paste0(collapse = ' / ') %>%
  { paste0('age (', paste0(., collapse = ' / '), ')') }
.venn_genotype_name <- venn_plot_data %>%
  filter(genotype) %>%
  group_by(genotype_regulation) %>%
  summarise(n = n()) %>%
  arrange(desc(genotype_regulation)) %>%
  mutate(name = paste(genotype_regulation, n)) %>%
  pull(name) %>%
  { paste0('genotype (', paste0(., collapse = ' / '), ')') }

ggplot() +
  geom_text(
    mapping = aes(x = 0, y = 1.8),
    label = tiss,
    size = 6
  ) +
  geom_venn(
    aes_string(A = 'age', B = 'genotype'),
    data = venn_plot_data,
    fill_color = c('#984EA3FF', '#FF7F00FF'),
    set_names = c(.venn_age_name, .venn_genotype_name),
    set_name_size = 4,
    auto_scale = TRUE,
    show_percentage = FALSE,
    show_outside = 'none',
  ) +
  labs(
    caption = paste('Number of genes regulated by age or genotype in', tiss, '; FDR <', th)
  ) +
  coord_fixed() +
  theme_void()
ggsave(
  paste0('results/venn-', tiss, '-', th, '.svg'),
  bg = 'white',
  width = 9,
  height = 6
)
```

```{r venn_10_702}
th <- 0.1
venn_genotype_10_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'genotype_10') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    gene_type = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_type,
    .before = everything()
  ) %>%
  filter(gene_type == 'protein_coding') %>%
  select(-row_index)

venn_genotype_70_genes <-
  edger_model(x, sample_info_groups, tiss, contrast = 'genotype_70') %>%
  mutate(
    gene_name = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_name,
    gene_stable_id = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_stable_id,
    gene_type = counts_cerebrum_raw[as.numeric(.$row_index), ]$gene_type,
    .before = everything()
  ) %>%
  filter(gene_type == 'protein_coding') %>%
  select(-row_index)


venn_plot_data <-
  full_join(
    venn_genotype_10_genes,
    venn_genotype_70_genes,
    by = c('gene_name', 'gene_stable_id'),
    suffix = c('.genotype_10', '.genotype_70')
  ) %>%
  mutate(
    genotype_10 = FDR.genotype_10 < th,
    genotype_70 = FDR.genotype_70 < th,
    genotype_10_regulation = if_else(logFC.genotype_10 < 0, 'up', 'down'),
    genotype_70_regulation = if_else(logFC.genotype_70 < 0, 'up', 'down'),
  )

.venn_genotype_10_name <- venn_plot_data %>%
  filter(genotype_10) %>%
  group_by(genotype_10_regulation) %>%
  summarise(n = n()) %>%
  arrange(desc(genotype_10_regulation)) %>%
  mutate(name = paste(genotype_10_regulation, n)) %>%
  pull(name) %>%
  paste0(collapse = ' / ') %>%
  { paste0('genotype 10 days (', paste0(., collapse = ' / '), ')') }
.venn_genotype_70_name <- venn_plot_data %>%
  filter(genotype_70) %>%
  group_by(genotype_70_regulation) %>%
  summarise(n = n()) %>%
  arrange(desc(genotype_70_regulation)) %>%
  mutate(name = paste(genotype_70_regulation, n)) %>%
  pull(name) %>%
  { paste0('genotype 10 weeks (', paste0(., collapse = ' / '), ')') }

ggplot() +
  geom_text(
    mapping = aes(x = 0, y = 1.8),
    label = tiss,
    size = 6
  ) +
  geom_venn(
    aes_string(A = 'genotype_10', B = 'genotype_70'),
    data = venn_plot_data,
    fill_color = c('#984EA3FF', '#FF7F00FF'),
    set_names = c(.venn_genotype_10_name, .venn_genotype_70_name),
    set_name_size = 3,
    auto_scale = TRUE,
    show_percentage = FALSE,
    show_outside = 'none',
  ) +
  labs(
    caption = paste('Number of genes regulated by genotype in 10 or 70 days old mice in', tiss, '; FDR <', th)
  ) +
  coord_fixed() +
  theme_void()
ggsave(
  paste0('results/venn-', tiss, '-in-age-', th, '.svg'),
  bg = 'white',
  width = 9,
  height = 6
)
```
