---
title: "portsmouth-darek-mdxbrain"
output:
  html_document:
    toc: yes
    df_print: paged

date: "_Last draft: `r format(Sys.time(), '%d %B, %Y %H:%M:%S')`_"
description: "mdx mice analysis"

knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_file = paste0(
        xfun::sans_ext(input), '-', format(Sys.time(), "%Y-%m-%d_%H-%M"), '.html'
      ),
      knit_root_dir = xfun::proj_root(),
      envir = globalenv()
    )
  })
---


```{r install_libraries, eval=FALSE, include=FALSE}
library(devtools)
devtools::install_github("zhangyuqing/sva-devel")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
# https://github.com/Bioconductor/bioconductor_docker/issues/22
BiocManager::install("preprocessCore", configure.args="--disable-threading")
```

```{r load_libraries, include=FALSE}
library(tidyverse)
library(preprocessCore)
library(RColorBrewer)

source('analysis/utils.R')
library(edgeR)
```

```{r read_data}
sample_info_rds_path <- 'data/sample_info.RDS'
sample_info <- readRDS(sample_info_rds_path)

counts_cerebrum_raw <- read_tsv('raw/cerebrum_gene-level-counts-annotated.tsv')
counts_cerebellum_raw <- read_tsv('raw/cerebellum_gene-level-counts-annotated.tsv')
counts_raw <-
  full_join(
    counts_cerebrum_raw,
    counts_cerebellum_raw,
    by = c(
      'Gene stable ID', 'Transcript stable ID', 'Exon stable ID',
      'Gene description', 'Chromosome/scaffold name', 'Gene start (bp)',
      'Gene end (bp)', 'Gene name', 'Gene type', 'Gene Synonym'
    )
  ) %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  )
```


```{r prepare_data}
x <- select(counts_raw, matches('sample_'))

mdx_groups <-
  sample_info %>%
  mutate(group = paste(genotype, age_days, tissue, sep = '_')) %>%
  `[`(match(colnames(x), .$id), ) %>%
  pull(group) %>%
  factor()

y <- DGEList(counts = x, group = mdx_groups)
keep_counts <- filterByExpr(y)
y <- y[keep_counts, , keep.lib.sizes = FALSE]


counts_normalized_log <-
  counts_raw %>%
  select(starts_with('sample')) %>%
  as.matrix() %>%
  # ComBat_seq(
  #   batch = c(rep(0, 20), rep(1, 20)),
  #   covar_mod = cbind(cov1, cov2),
  # ) %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}
counts <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_normalized_log) # %>%
  # mutate(
  #   n_outliers = apply(
  #     select(., starts_with('sample_')),
  #     1,
  #     function(x) {length(boxplot.stats(x)$out)}
  #   ),
  # ) %>%
  # filter(
  #   n_outliers < 1,
  #   gene_type == 'protein_coding',
  #   pmax(
  #     rowMeans(across(matches('^sample_BD\\d$'))),
  #     rowMeans(across(matches('^sample_BD\\dC$'))),
  #     rowMeans(across(matches('^sample_BW\\d$'))),
  #     rowMeans(across(matches('^sample_BW\\dC$'))),
  #     rowMeans(across(matches('^sample_MD\\d$'))),
  #     rowMeans(across(matches('^sample_MD\\dC$'))),
  #     rowMeans(across(matches('^sample_MW\\d$'))),
  #     rowMeans(across(matches('^sample_MW\\dC$')))
  #   ) > 3,
  # )
```

```{r}
plot_info <-
  sample_info %>%
  arrange(desc(genotype), desc(tissue), age_days) %>%
  mutate(
    `10 days / 70 days` = if_else(age_days == 10, 'green', 'magenta'),
    `cerebrum / cerebellum` = if_else(tissue == 'cerebrum', 'red', 'blue'),
    `wt / mdx` = if_else(genotype == 'wt', 'yellow', 'black'),
  )
```


```{r fit_model}
y <- calcNormFactors(y)
design <- model.matrix(~ 0 + mdx_groups)
y <- estimateDisp(y, design)

fit <- glmQLFit(y, design)
```

```{r}
my.contrasts <- makeContrasts(
  #overall genotype effect
  genotype=
    (
      mdx_groupswt_10_cerebrum + mdx_groupswt_70_cerebrum
      + mdx_groupswt_10_cerebellum + mdx_groupswt_70_cerebellum
    )
    - (
      mdx_groupsmdx_10_cerebrum + mdx_groupsmdx_70_cerebrum
      + mdx_groupsmdx_10_cerebellum + mdx_groupsmdx_70_cerebellum
    ),
  genotype_70_cerebrum=((mdx_groupsmdx_70_cerebrum - mdx_groupswt_70_cerebrum)),
  genotype_70_cerebellum=((mdx_groupsmdx_70_cerebellum - mdx_groupswt_70_cerebellum)),
  genotype_10_cerebrum=((mdx_groupsmdx_10_cerebrum - mdx_groupswt_10_cerebrum)),
  genotype_10_cerebellum=((mdx_groupsmdx_10_cerebellum - mdx_groupswt_10_cerebellum)),
  # genotype_age_days=((mdx_groupsmdx_70_cerebellum - mdx_groupsmdx_10_cerebellum) - (mdx_groupswt_70_cerebellum - mdx_groupswt_10_cerebellum)),
  # genotype_tissue=((mdx_groupsmdx_10_cerebellum - mdx_groupsmdx_10_cerebrum) - (mdx_groupswt_10_cerebellum - mdx_groupswt_10_cerebrum)),
  # genotype_age_days=((mdx_groupsmdx_70_cerebellum + mdx_groupsmdx_10_cerebellum) - (mdx_groupswt_70_cerebellum + mdx_groupswt_10_cerebellum)),
  # genotype_tissue=((mdx_groupsmdx_10_cerebellum + mdx_groupsmdx_10_cerebrum) - (mdx_groupswt_10_cerebellum + mdx_groupswt_10_cerebrum)),
  levels = design
)
genotype <- glmQLFTest(fit, contrast=my.contrasts[,"genotype"])
genotype_70_cerebrum <- glmQLFTest(fit, contrast=my.contrasts[,"genotype_70_cerebrum"])
genotype_70_cerebellum <- glmQLFTest(fit, contrast=my.contrasts[,"genotype_70_cerebellum"])
genotype_10_cerebrum <- glmQLFTest(fit, contrast=my.contrasts[,"genotype_10_cerebrum"])
genotype_10_cerebellum <- glmQLFTest(fit, contrast=my.contrasts[,"genotype_10_cerebellum"])
# genotype_age_days <- glmQLFTest(fit, contrast=my.contrasts[,"genotype_age_days"])
# genotype_tissue <- glmQLFTest(fit, contrast=my.contrasts[,"genotype_tissue"])
z <- topTags(genotype, n = 'inf')
z1 <- topTags(genotype_70_cerebrum, n = 'inf')
z2 <- topTags(genotype_70_cerebellum, n = 'inf')
z3 <- topTags(genotype_10_cerebrum, n = 'inf')
z4 <- topTags(genotype_10_cerebellum, n = 'inf')
z5 <- topTags(genotype_age_days, n = 'inf')
z6 <- topTags(genotype_tissue, n = 'inf')

g <- bind_rows(
  z$table %>% as_tibble(rownames = 'a'),
  # z1$table %>% as_tibble(rownames = 'a'),
  # z2$table %>% as_tibble(rownames = 'a'),
  # z3$table %>% as_tibble(rownames = 'a'),
  # z4$table %>% as_tibble(rownames = 'a')
  # z5$table %>% as_tibble(rownames = 'a'),
  # z6$table %>% as_tibble(rownames = 'a')
) %>%
arrange(FDR)
```

```{r draw_heatmap_03, echo=FALSE, fig.align="center", fig.height=10, fig.width=10, results='asis', include=TRUE, warning=FALSE}
zz <-
  g %>%
  filter(FDR < 0.001) %>%
  pull(a) %>%
  unique()

plot_data <-
  counts %>%
  `[`(as.numeric(zz), )

.x <- plot_data %>% select(all_of(plot_info$id)) %>% as.matrix()
rownames(.x) <- plot_data$gene_name
.rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
.col_side_colors <- plot_info %>% select(contains('/')) %>% as.matrix()
.way = 'genotype / genotype:tissue / genotype:age_days / genotype:tissue:age_days'

# png('results/genotype-edger.png', width = 1000*1.5, height = 1600*1.5, pointsize = 19)
h <- heatmap.3(
    x = .x,
    main = 'edgeR',
    Rowv = TRUE,
    Colv = FALSE,
    dendrogram = 'row',
    distfun = function(x) as.dist(1-cor(t(x))),
    scale = "row",
    breaks = seq(-2.5, 2.5, length.out = 25),
    col = rev(.rdbu_ramp(24)),
    trace = "none",
    ColSideColors = .col_side_colors,
    ColSideColorsSize = 6,
    key = FALSE,
    lhei = c(1.5, 20),
    lwid = c(1,5),
    cexRow = 1.2,
    cexCol = 1.5,
    margins = c(9, 8),
)
# dev.off()
```

```{r}
ct <- cutree(as.hclust(h$rowDendrogram), k = 2)
unique(ct[h$rowInd])


tibble(cluster = ct, gene_name = names(ct)) %>%
  filter(cluster == 2) %>%
  pull(gene_name) %>%
  paste(collapse = '\n') %>%
  cat

# ct %>% as.data.frame %>% rownames_to_column('gene_name') %>%  write_csv2('edger-clusters.csv')
```
