---
title: "portsmouth-darek-mdxbrain"
output:
  html_document:
    toc: yes
    df_print: paged

date: "_Last draft: `r format(Sys.time(), '%d %B, %Y %H:%M:%S')`_"
description: "mdx mice analysis"

knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_file = paste0(
        xfun::sans_ext(input), '-', format(Sys.time(), "%Y-%m-%d_%H-%M"), '.html'
      ),
      knit_root_dir = xfun::proj_root(),
      envir = globalenv()
    )
  })
---


```{r install_libraries, eval=FALSE, include=FALSE}
library(devtools)
devtools::install_github("zhangyuqing/sva-devel")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
# https://github.com/Bioconductor/bioconductor_docker/issues/22
BiocManager::install("preprocessCore", configure.args="--disable-threading")
```

```{r load_libraries, include=FALSE}
library(tidyverse)
library(preprocessCore)
library(future.apply)
library(rstatix)
library(gplots)
library(devtools)
source_url("https://raw.githubusercontent.com/obigriffith/biostar-tutorials/master/Heatmaps/heatmap.3.R")
library(RColorBrewer)
library(tictoc)
library(sva)

gg_qqplot <- function(ps, ci = 0.95) {
  n  <- length(ps)
  df <- data.frame(
    observed = -log10(sort(ps)),
    expected = -log10(ppoints(n)),
    clower   = -log10(qbeta(p = (1 - ci) / 2, shape1 = 1:n, shape2 = n:1)),
    cupper   = -log10(qbeta(p = (1 + ci) / 2, shape1 = 1:n, shape2 = n:1))
  )
  log10Pe <- expression(paste("Expected -log"[10], plain(P)))
  log10Po <- expression(paste("Observed -log"[10], plain(P)))
  ggplot(df) +
    geom_ribbon(
      mapping = aes(x = expected, ymin = clower, ymax = cupper),
      alpha = 0.1
    ) +
    geom_point(aes(expected, observed), shape = 16, size = 1) +
    geom_abline(intercept = 0, slope = 1, alpha = 0.5) +
    geom_line(aes(expected, cupper), linetype = 2, size = 0.5) +
    geom_line(aes(expected, clower), linetype = 2, size = 0.5) +
    xlab(log10Pe) +
    ylab(log10Po)
}

inflation <- function(ps) {
  chisq <- qchisq(1 - ps, 1)
  lambda <- median(chisq) / qchisq(0.5, 1)
  lambda
}
```

```{r read_data, include=FALSE}
counts_cerebrum_raw <- read_tsv('raw/cerebrum_gene-level-counts-annotated.tsv')
counts_cerebellum_raw <- read_tsv('raw/cerebellum_gene-level-counts-annotated.tsv')

counts_raw <-
  full_join(
    counts_cerebrum_raw,
    counts_cerebellum_raw,
    by = c(
      'Gene stable ID', 'Transcript stable ID', 'Exon stable ID',
      'Gene description', 'Chromosome/scaffold name', 'Gene start (bp)',
      'Gene end (bp)', 'Gene name', 'Gene type', 'Gene Synonym'
    )
  ) %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  ) %>%
  filter(rowMeans(across(contains('sample'))) > 5)

sample_pattern <- "(sample_(\\w)(\\w)\\d)(C?)"
sample_info <-
  tibble(id = colnames(counts_raw)) %>%
  mutate(mouse_id = str_match(id, sample_pattern)[, 2]) %>%
  filter(!is.na(mouse_id)) %>%
  mutate(
    genotype = str_match(id, sample_pattern)[,3],
    age_days = str_match(id, sample_pattern)[,4],
    tissue = str_match(id, sample_pattern)[,5],
  ) %>%
  mutate(
    genotype = if_else(
      genotype == 'B',
      'wt',
      if_else(genotype == 'M', 'mdx', NULL)
    ),
    age_days = if_else(
      age_days == 'D',
      10,
      if_else(age_days == 'W', 70, NULL)
    ),
    tissue = if_else(
      tissue == 'C',
      'cerebrum',
      if_else(tissue == '', 'cerebellum', NULL)
    ),
  )
```

### Normalization
```{r normalize_log_counts}
counts_normalized_log <-
  counts_raw %>%
  select(starts_with('sample')) %>%
  as.matrix() %>%
  ComBat_seq(
    batch = c(rep(0, 20), rep(1, 20)),
  ) %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}

counts <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_normalized_log)
```

```{r write_counts, eval=FALSE}
counts %>% write_csv2('results/adj_norm_counts_excel.csv')
```

### Analysis
```{r prepare_analysis, include=FALSE}
compute_anova <- function(x, template) {
  template %>%
    left_join(as_tibble(x, rownames = 'id'), by = 'id') %>%
    anova_test(
      dv = value,
      wid = mouse_id,
      between = c(genotype, age_days),
      within = tissue
    ) %>%
    get_anova_table() %>%
    data.frame() %>%
    select(Effect, p) %>%
    pivot_wider(names_from = Effect, values_from = p)
}
```

```{r compute_anova, include=FALSE}
anova_rds_path <- 'data/mdx_3way_anova_002.RDS'
if (!file.exists(anova_rds_path)) {
  if (Sys.info()['sysname'] == 'Windows') {
    plan(multisession)
  } else {
    # Rscript -e "
    #   rmarkdown::render(
    #     'analysis/main.Rmd',
    #     knit_root_dir = xfun::proj_root(),
    #     output_file = base::paste0(
    #       xfun::sans_ext('main.Rmd'), '-',
    #       base::format(Sys.time(), \"%Y-%m-%d_%H-%M\"), '.html'
    #     )
    #   )
    # "
    plan(multicore)
  }

  mdx_anova <- future_apply(
    counts %>% select(contains('sample')),
    1,
    compute_anova,
    template = sample_info
  )

  bind_rows(mdx_anova) %>%
    mutate(
      across(
        c(
          genotype, age_days, tissue, `genotype:age_days`, `genotype:tissue`,
          `age_days:tissue`, `genotype:age_days:tissue`
        ),
        list(fdr = p.adjust),
        method = 'fdr'
      )
    ) %>%
    {{
      bind_cols(
        counts %>% select(!contains('sample')),
        .
      )
    }} %>%
    left_join(
      counts,
      by = c(
        'gene_stable_id', 'gene_description', 'chr', 'gene_name',
        'gene_type', 'gene_synonym'
      )
    ) %>%
    saveRDS(anova_rds_path)
}
```

```{r write_anova_results, eval=FALSE}
mdx_anova %>% write_csv('results/anova_results.csv')
mdx_anova %>% write_csv2('results/anova_results_excel.csv')
```

```{r}
mdx_anova <- readRDS(anova_rds_path)
```

```{r draw_heatmaps, echo=FALSE, fig.align="center", fig.height=10, fig.width=13, results='asis', include=TRUE, warning=FALSE}
groups <- c(
  'genotype_fdr', 'genotype:tissue_fdr', 'genotype:age_days_fdr',
  'genotype:age_days:tissue_fdr'
)
for (way in groups) {
  plot_data <-
    mdx_anova %>%
    arrange(.data[[sub('_fdr', '', way)]]) %>%
    slice(1:50)

  row_5 <- plot_data[5, way] %>% unlist() %>% unname()
  row_50 <- plot_data[50, way] %>% unlist() %>% unname()
  cat(paste0("factor: ", way, '\n\n'))
  cat(paste0("threshold: ", row_50), '\n\n')
  cat(paste0("top5 threshold: ", row_5), '(', pull(plot_data, gene_name)[1:5], ')')

  .x <- plot_data %>% select(contains('sample')) %>% as.matrix()
  .rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
  .col_side_colors <- as.matrix(tibble(
      '10 days / 70 days' = rep(c(rep('green', 5), rep('magenta', 5)), 4),
      'wt / mdx' = rep(c(rep('yellow', 10), rep('black', 10)), 2),
      'cerebrum / cerebellum' = c(rep('red', 20), rep('blue', 20)),
  ))
  heatmap.3(
      x = .x,
      Rowv = TRUE,
      Colv = FALSE,
      dendrogram = 'row',
      distfun = function(x) as.dist(1-cor(t(x))),
      scale = "row",
      col = rev(.rdbu_ramp(24)),
      trace = "none",
      ColSideColors = .col_side_colors,
      ColSideColorsSize = 8,
      key = FALSE,
      lhei = c(1.3, 20),
      labRow = pull(plot_data, gene_name),
      cexRow = 1.2,
      cexCol = 1.5,
      offsetRow = 0,
  )
}
```
