---
title: "portsmouth-darek-mdxbrain"
output:
  html_document:
    toc: yes
    df_print: paged

date: "_Last draft: `r format(Sys.time(), '%d %B, %Y %H:%M:%S')`_"
description: "mdx mice analysis"

knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_file = paste0(
        xfun::sans_ext(input), '-', format(Sys.time(), "%Y-%m-%d_%H-%M"), '.html'
      ),
      knit_root_dir = xfun::proj_root(),
      envir = globalenv()
    )
  })
---


```{r install_libraries, eval=FALSE, include=FALSE}
library(devtools)
devtools::install_github("zhangyuqing/sva-devel")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
# https://github.com/Bioconductor/bioconductor_docker/issues/22
BiocManager::install("preprocessCore", configure.args="--disable-threading")
```

```{r load_libraries, include=FALSE}
library(tidyverse)
library(preprocessCore)
library(future.apply)
library(rstatix)
library(RColorBrewer)
library(tictoc)
library(sva)
library(knitr)

source('analysis/utils.R')
```

```{r read_data, include=FALSE}
anova_rds_path <- 'data/mdx_3way_anova_010.RDS'

counts_cerebrum_raw <- read_tsv('raw/cerebrum_gene-level-counts-annotated.tsv')
counts_cerebellum_raw <- read_tsv('raw/cerebellum_gene-level-counts-annotated.tsv')

counts_raw <-
  full_join(
    counts_cerebrum_raw,
    counts_cerebellum_raw,
    by = c(
      'Gene stable ID', 'Transcript stable ID', 'Exon stable ID',
      'Gene description', 'Chromosome/scaffold name', 'Gene start (bp)',
      'Gene end (bp)', 'Gene name', 'Gene type', 'Gene Synonym'
    )
  ) %>%
  rename(
    gene_stable_id = `Gene stable ID`,
    gene_description = `Gene description`,
    chr = `Chromosome/scaffold name`,
    gene_name = `Gene name`,
    gene_type = `Gene type`,
    gene_synonym = `Gene Synonym`
  ) %>%
  filter(rowMeans(across(contains('sample'))) > 1)

sample_pattern <- "(sample_(\\w)(\\w)\\d)(C?)"
sample_info <-
  tibble(id = colnames(counts_raw)) %>%
  mutate(mouse_id = str_match(id, sample_pattern)[, 2]) %>%
  filter(!is.na(mouse_id)) %>%
  mutate(
    genotype = str_match(id, sample_pattern)[,3],
    age_days = str_match(id, sample_pattern)[,4],
    tissue = str_match(id, sample_pattern)[,5],
  ) %>%
  mutate(
    genotype = if_else(
      genotype == 'B',
      'wt',
      if_else(genotype == 'M', 'mdx', NULL)
    ),
    age_days = if_else(
      age_days == 'D',
      10,
      if_else(age_days == 'W', 70, NULL)
    ),
    tissue = if_else(
      tissue == 'C',
      'cerebrum',
      if_else(tissue == '', 'cerebellum', NULL)
    ),
  )
```

### Normalization
```{r normalize_log_counts}
cov1 <- rep(c(rep(0, 5), rep(1, 5)), 4)
cov2 <- rep(c(rep(0, 10), rep(1, 10)), 2)

counts_normalized_log <-
  counts_raw %>%
  select(starts_with('sample')) %>%
  as.matrix() %>%
  ComBat_seq(
    batch = c(rep(0, 20), rep(1, 20)),
    covar_mod = cbind(cov1, cov2),
  ) %>%
  normalize.quantiles(keep.names = TRUE) %>%
  {{ log2(. + 1) }}

counts <-
  counts_raw %>%
  select(
    gene_stable_id, gene_description, chr, gene_name, gene_type, gene_synonym
  ) %>%
  bind_cols(counts_normalized_log)
```

```{r write_counts, eval=FALSE}
counts %>% write_csv2('results/adj_norm_counts_excel.csv')
counts %>% write_csv('results/data.csv')
```

### Analysis
```{r prepare_analysis, include=FALSE}
compute_anova <- function(x, template) {
  template %>%
    left_join(as_tibble(x, rownames = 'id'), by = 'id') %>%
    anova_test(
      dv = value,
      wid = mouse_id,
      between = c(genotype, age_days),
      within = tissue
    ) %>%
    get_anova_table() %>%
    data.frame() %>%
    select(Effect, p) %>%
    pivot_wider(names_from = Effect, values_from = p)
}
```

```{r compute_anova, include=FALSE}
if (Sys.info()['sysname'] == 'Windows') {
  plan(multisession)
} else {
  # Rscript -e "
  #   rmarkdown::render(
  #     'analysis/main.Rmd',
  #     knit_root_dir = xfun::proj_root(),
  #     output_file = base::paste0(
  #       xfun::sans_ext('main.Rmd'), '-',
  #       base::format(Sys.time(), \"%Y-%m-%d_%H-%M\"), '.html'
  #     )
  #   )
  # "
  plan(multicore)
}
if (!file.exists(anova_rds_path)) {
  mdx_anova <- future_apply(
    counts %>% select(contains('sample')),
    1,
    compute_anova,
    template = sample_info
  )

  bind_rows(mdx_anova) %>%
    mutate(
      across(
        c(
          genotype, age_days, tissue, `genotype:age_days`, `genotype:tissue`,
          `age_days:tissue`, `genotype:age_days:tissue`
        ),
        list(fdr = p.adjust),
        method = 'fdr'
      )
    ) %>%
    {{
      bind_cols(
        counts %>% select(!contains('sample')),
        .
      )
    }} %>%
    left_join(
      counts,
      by = c(
        'gene_stable_id', 'gene_description', 'chr', 'gene_name',
        'gene_type', 'gene_synonym'
      )
    ) %>%
    saveRDS(anova_rds_path)
}
```

```{r write_anova_results, eval=FALSE}
mdx_anova %>% write_csv('results/anova_results.csv')
mdx_anova %>% write_csv2('results/anova_results_excel.csv')
```

```{r}
mdx_anova <- readRDS(anova_rds_path)
```

```{r draw_heatmaps, echo=FALSE, fig.align="center", fig.height=10, fig.width=13, results='asis', include=TRUE, warning=FALSE}
groups <- list(
  'genotype' = list('genotype_fdr', 0.1, 4),
  'genotype:tissue' = list('genotype:tissue_fdr', 0.0005, 3),
  'genotype:age_days' = list('genotype:age_days_fdr', 0.3, 4),
  'genotype:age_days:tissue' = list('genotype:age_days:tissue_fdr', 0.1, 3)
)
for (way in names(groups)) {
  group <- groups[[way]]
  way_fdr = group[[1]]
  th = group[[2]]
  k = group[[3]]

  plot_data <-
    mdx_anova %>%
    arrange(.data[[way]]) %>%
    filter(.data[[way_fdr]] < th)

  .x <- plot_data %>% select(contains('sample')) %>% as.matrix()
  .rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
  .col_side_colors <- as.matrix(tibble(
      '10 days / 70 days' = rep(c(rep('green', 5), rep('magenta', 5)), 4),
      'wt / mdx' = rep(c(rep('yellow', 10), rep('black', 10)), 2),
      'cerebrum / cerebellum' = c(rep('red', 20), rep('blue', 20)),
  ))
  print(paste(way, ' FDR < ', th, collapse = ''))
  h <- heatmap.3(
      x = .x,
      main = paste(way, ' FDR < ', th, collapse = ''),
      Rowv = TRUE,
      Colv = FALSE,
      dendrogram = 'row',
      distfun = function(x) as.dist(1-cor(t(x))),
      scale = "row",
      col = rev(.rdbu_ramp(24)),
      trace = "none",
      ColSideColors = .col_side_colors,
      ColSideColorsSize = 8,
      key = FALSE,
      lhei = c(1.5, 20),
      labRow = pull(plot_data, gene_name),
      cexRow = 1.2,
      cexCol = 1.5,
      # offsetRow = 0,
  )
  ct <- cutree(as.hclust(h$rowDendrogram), k = k)
  plot_data %>%
    mutate(ct = ct, .after = gene_name) %>%
    select(gene_stable_id, gene_description, gene_type, gene_name, ct) %>%
    `[`(rev(h$rowInd), TRUE) %>%
    select(-gene_description, -gene_type) %>%
    kable() %>%
    print()
}
```


### Post-hoc analysis

```{r compute_folds}
get_sample_names <- function(sample_info, genotype, age_days, tissue) {
  stopifnot(genotype %in% c('wt', 'mdx'))
  stopifnot(age_days %in% c(10, 70))
  stopifnot(tissue %in% c('cerebrum', 'cerebellum'))
  genotype_ = genotype
  age_days_ = age_days
  tissue_ = tissue

  sample_info %>%
    filter(
      genotype == genotype_,
      age_days == age_days_,
      tissue == tissue_
    ) %>%
    pull(id)
}

mdx_anova_folds <-
  mdx_anova %>%
  mutate(
    wt_10_cerebrum_mean = apply(
      select(., matches('sample_[MB][DW][1-5]C?')),
      1,
      function(x, names) {mean(x[names])},
      names = get_sample_names(sample_info, 'wt', 10, 'cerebrum')
    ),
    wt_10_cerebellum_mean = apply(
      select(., matches('sample_[MB][DW][1-5]C?')),
      1,
      function(x, names) {mean(x[names])},
      names = get_sample_names(sample_info, 'wt', 10, 'cerebellum')
    ),
    wt_70_cerebrum_mean = apply(
      select(., matches('sample_[MB][DW][1-5]C?')),
      1,
      function(x, names) {mean(x[names])},
      names = get_sample_names(sample_info, 'wt', 70, 'cerebrum')
    ),
    wt_70_cerebellum_mean = apply(
      select(., matches('sample_[MB][DW][1-5]C?')),
      1,
      function(x, names) {mean(x[names])},
      names = get_sample_names(sample_info, 'wt', 70, 'cerebellum')
    ),
    mdx_10_cerebrum_mean = apply(
      select(., matches('sample_[MB][DW][1-5]C?')),
      1,
      function(x, names) {mean(x[names])},
      names = get_sample_names(sample_info, 'mdx', 10, 'cerebrum')
    ),
    mdx_10_cerebellum_mean = apply(
      select(., matches('sample_[MB][DW][1-5]C?')),
      1,
      function(x, names) {mean(x[names])},
      names = get_sample_names(sample_info, 'mdx', 10, 'cerebellum')
    ),
    mdx_70_cerebrum_mean = apply(
      select(., matches('sample_[MB][DW][1-5]C?')),
      1,
      function(x, names) {mean(x[names])},
      names = get_sample_names(sample_info, 'mdx', 70, 'cerebrum')
    ),
    mdx_70_cerebellum_mean = apply(
      select(., matches('sample_[MB][DW][1-5]C?')),
      1,
      function(x, names) {mean(x[names])},
      names = get_sample_names(sample_info, 'mdx', 70, 'cerebellum')
    ),
  ) %>%
  mutate(
    fold_10_cerebrum =
      abs(wt_10_cerebrum_mean - mdx_10_cerebrum_mean),
    fold_10_cerebellum =
      abs(wt_10_cerebellum_mean - mdx_10_cerebellum_mean),
    fold_70_cerebrum =
      abs(wt_70_cerebrum_mean - mdx_70_cerebrum_mean),
    fold_70_cerebellum =
      abs(wt_70_cerebellum_mean - mdx_70_cerebellum_mean),
    fold_max =
      apply(
        select(., matches('(wt|mdx)_(10|70)_(cerebrum|cerebellum)_mean')), 1, max
      )
    - apply(
        select(., matches('(wt|mdx)_(10|70)_(cerebrum|cerebellum)_mean')), 1, min
      )
  )
```

```{r}
mdx_anova_folds %>%
  select(-starts_with('fold_'), -starts_with('sample_')) %>%
  write_csv('results/results_anova.csv')
```

```{r compute_post_hoc}
compute_ttest_genotype <- function(x, template) {
  template %>%
  left_join(as_tibble(x, rownames = 'id'), by = 'id') %>%
  # mutate(
  #   genotype = as.factor(genotype),
  #   age_days = as.factor(age_days),
  #   tissue = as.factor(tissue)
  # ) %>%
  group_by(age_days, tissue) %>%
  pairwise_t_test(
      value ~ genotype,
      pool.sd = FALSE,
      var.equal = TRUE
  ) %>%
  select(age_days, tissue, p)
}

mdx_anova_df <- as.data.frame(mdx_anova_folds)
rownames(mdx_anova_df) <- mdx_anova_df$gene_stable_id
print(mdx_anova_df %>% filter(genotype_fdr < 0.1) %>% dim)
ttest_genotype_bonf <- future_apply(
    mdx_anova_df %>% filter(genotype_fdr < 0.1) %>% select(contains('sample')),
    1,
    compute_ttest_genotype,
    template = sample_info
)
ttest_genotype_bonf <-
  bind_rows(ttest_genotype_bonf, .id = 'gene_stable_id') %>%
  group_by(gene_stable_id) %>%
  mutate(p_bf = p.adjust(p, 'bonferroni')) %>%
  ungroup()

fold_th <- 0
post_hoc_genes <-
  ttest_genotype_bonf %>%
  left_join(
    mdx_anova_folds %>% select(gene_stable_id, gene_name, starts_with('fold_')),
    by = 'gene_stable_id'
  ) %>%
  filter(
    age_days == 10 & tissue == 'cerebrum' & fold_10_cerebrum > fold_th
    | age_days == 10 & tissue == 'cerebellum' & fold_10_cerebellum > fold_th
    | age_days == 70 & tissue == 'cerebrum' & fold_70_cerebrum > fold_th
    | age_days == 70 & tissue == 'cerebellum' & fold_70_cerebellum > fold_th
  ) %>%
  group_by(age_days, tissue) %>%
  filter(p_bf < 0.01)
post_hoc_genes %>%
  summarize(n = n(), .groups = 'drop')

a <- 10
t <- 'cerebrum'
print(c(a, t))
d1 <- post_hoc_genes %>%
  filter(age_days == a, tissue == t) %>%
  pull(gene_name)
print(sort(d1))

a <- 10
t <- 'cerebellum'
print(c(a, t))
d2 <- post_hoc_genes %>%
  filter(age_days == a, tissue == t) %>%
  pull(gene_name)
print(sort(d2))

a <- 70
t <- 'cerebrum'
print(c(a, t))
d3 <- post_hoc_genes %>%
  filter(age_days == a, tissue == t) %>%
  pull(gene_name)
print(sort(d3))

a <- 70
t <- 'cerebellum'
print(c(a, t))
d4 <- post_hoc_genes %>%
  filter(age_days == a, tissue == t) %>%
  pull(gene_name)
print(sort(d4))

print("common genes: ")
discard(intersect(d1, d2), is.na)
discard(intersect(d1, d3), is.na)
discard(intersect(d1, d4), is.na)
discard(intersect(d2, d3), is.na)
discard(intersect(d2, d4), is.na)
discard(intersect(d3, d4), is.na)
```

```{r draw_heatmaps_genotype, echo=FALSE, fig.align="center", fig.height=12, fig.width=13, results='asis', include=TRUE, warning=FALSE}
plot_data <-
  mdx_anova_folds %>%
  filter(genotype_fdr < 0.1) %>%
  left_join(ttest_genotype_bonf_wide, by = 'gene_stable_id') %>%
  filter(
    (`wt vs mdx in 10_cerebrum p_bf` < 0.05 & abs(fold_10_cerebrum) > 1)
    | (`wt vs mdx in 10_cerebellum p_bf` < 0.05 & abs(fold_10_cerebellum) > 1)
    | (`wt vs mdx in 70_cerebrum p_bf` < 0.05 & abs(fold_70_cerebrum) > 1)
    | (`wt vs mdx in 70_cerebellum p_bf` < 0.05 & abs(fold_70_cerebellum) > 1)
  )

.x <- plot_data %>% select(contains('sample')) %>% as.matrix()
.rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
.col_side_colors <- as.matrix(tibble(
    '10 days / 70 days' = rep(c(rep('green', 5), rep('magenta', 5)), 4),
    'wt / mdx' = rep(c(rep('yellow', 10), rep('black', 10)), 2),
    'cerebrum / cerebellum' = c(rep('red', 20), rep('blue', 20)),
))
h <- heatmap.3(
    x = .x,
    Rowv = TRUE,
    Colv = FALSE,
    dendrogram = 'row',
    distfun = function(x) as.dist(1-cor(t(x))),
    scale = "row",
    col = rev(.rdbu_ramp(24)),
    trace = "none",
    ColSideColors = .col_side_colors,
    ColSideColorsSize = 8,
    key = FALSE,
    lhei = c(1.5, 20),
    labRow = pull(plot_data, gene_name),
    cexRow = 1.2,
    cexCol = 1.5,
    # offsetRow = 0,
)

  ct <- cutree(as.hclust(h$rowDendrogram), k = 3)
  plot_clusters <-
    plot_data %>%
    mutate(ct = ct, .after = gene_name) %>%
    `[`(rev(h$rowInd), TRUE) %>%
    select(gene_stable_id, ct)
```

```{r save_post_hoc, eval=FALSE}
ttest_genotype_bonf_wide <- ttest_genotype_bonf %>%
  pivot_wider(
    names_from = c(age_days, tissue),
    values_from = c(p, p_bf),
    names_vary = 'slowest',
    names_glue = 'wt vs mdx in {age_days}_{tissue} {.value}'
  )

mdx_anova_folds %>%
  select(-contains('sample_')) %>%
  filter(genotype_fdr < 0.1) %>%
  left_join(ttest_genotype_bonf_wide, by = 'gene_stable_id') %>%
  select(-genotype:-`genotype:age_days:tissue_fdr`, -ends_with('_mean')) %>%
  select(-fold_max) %>%
  left_join(plot_clusters, by = 'gene_stable_id') %>%
  relocate(ct, .after = gene_name) %>%
  write_csv('results/results_genotype_posthost.csv')
```


#### fdr < 0.01, heatmap
```{r draw_heatmaps2, echo=FALSE, fig.align="center", fig.height=10, fig.width=13, results='asis', include=TRUE, warning=FALSE}
bonf_th <- 0.01
fold_th <- 0
groups <- list(
  c(10, 'cerebrum'), c(10, 'cerebellum'), c(70, 'cerebrum'), c(70, 'cerebellum')
)

for (group in groups) {
  print(paste0('##### age_days: ', group[1], ', tissue: ', group[2]))
  post_hoc_genes <-
    ttest_genotype_bonf %>%
    filter(age_days == group[1], tissue == group[2]) %>%
    filter(p_bf < bonf_th) %>%
    pull(gene_stable_id)

  plot_data <-
    mdx_anova_folds %>%
    filter(.data[[paste0('fold_', group[1], '_', group[2])]] > fold_th) %>%
    filter(gene_stable_id %in% post_hoc_genes)

  .x <- plot_data %>% select(contains('sample')) %>% as.matrix()
  .rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
  .col_side_colors <- as.matrix(tibble(
      '10 days / 70 days' = rep(c(rep('green', 5), rep('magenta', 5)), 4),
      'wt / mdx' = rep(c(rep('yellow', 10), rep('black', 10)), 2),
      'cerebrum / cerebellum' = c(rep('red', 20), rep('blue', 20)),
  ))
  heatmap.3(
      x = .x,
      Rowv = TRUE,
      Colv = FALSE,
      dendrogram = 'row',
      distfun = function(x) as.dist(1-cor(t(x))),
      scale = "row",
      col = rev(.rdbu_ramp(24)),
      trace = "none",
      ColSideColors = .col_side_colors,
      ColSideColorsSize = 8,
      key = FALSE,
      lhei = c(1.3, 20),
      labRow = pull(plot_data, gene_name),
      cexRow = 1.2,
      cexCol = 1.5,
      # offsetRow = 0,
  )
}
```

#### fdr < 0.01, fold 2, heatmap
```{r draw_heatmaps3, echo=FALSE, fig.align="center", fig.height=10, fig.width=13, results='asis', include=TRUE, warning=FALSE}
bonf_th <- 0.01
fold_th <- 1

for (group in groups){
  print(paste0('##### age_days: ', group[1], ', tissue: ', group[2]))
  post_hoc_genes <-
    ttest_genotype_bonf %>%
    filter(age_days == group[1], tissue == group[2]) %>%
    filter(p_bf < bonf_th) %>%
    pull(gene_stable_id)

  plot_data <-
    mdx_anova_folds %>%
    filter(.data[[paste0('fold_', group[1], '_', group[2])]] > fold_th) %>%
    filter(gene_stable_id %in% post_hoc_genes)

  .x <- plot_data %>% select(contains('sample')) %>% as.matrix()
  .rdbu_ramp <- colorRampPalette(brewer.pal(11, "RdBu"))
  .col_side_colors <- as.matrix(tibble(
      '10 days / 70 days' = rep(c(rep('green', 5), rep('magenta', 5)), 4),
      'wt / mdx' = rep(c(rep('yellow', 10), rep('black', 10)), 2),
      'cerebrum / cerebellum' = c(rep('red', 20), rep('blue', 20)),
  ))
  heatmap.3(
      x = .x,
      Rowv = TRUE,
      Colv = FALSE,
      dendrogram = 'row',
      distfun = function(x) as.dist(1-cor(t(x))),
      scale = "row",
      col = rev(.rdbu_ramp(24)),
      trace = "none",
      ColSideColors = .col_side_colors,
      ColSideColorsSize = 8,
      key = FALSE,
      lhei = c(1.3, 20),
      labRow = pull(plot_data, gene_name),
      cexRow = 1.2,
      cexCol = 1.5,
      # offsetRow = 0,
  )
}
```
